kolibri/core/assets/src/heartbeat.js

```diff
@@ -0,0 +1,58 @@
+const logging = require('kolibri.lib.logging').getLogger(__filename);
+
+class HeartBeat {
+  constructor(kolibri) {
+    this.kolibri = kolibri;
+    // Do this to have a consistent callback that has 'this' properly bound
+    // but can be repeatedly referenced to add and remove listeners.
+    this.setActiveCallback = this.setActive.bind(this);
+    this.setInactive();
+    this.start();
+  }
+  start() {
+    logging.debug('Starting heartbeat');
+    this.setActivityListeners();
+    this.beat();
+  }
+  setActivityListeners() {
+    this.events.forEach((event) => {
+      document.addEventListener(event, this.setActiveCallback, true);
+    });
+  }
+  clearActivityListeners() {
+    this.events.forEach((event) => {
+      document.removeEventListener(event, this.setActiveCallback, true);
+    });
+  }
+  setActive() {
+    this.active = true;
```

Comment:
In my (admittedly not very scientific) profiling of this versus the alternative, it reduces the total time spent executing the heartbeat functions (not including the actual polling) by a bit more than half. Admittedly, this is half of 1.5ms on my machine, but, then again, I also have a fairly performant machine. So, I think it may be a small optimization, but one that is worth doing anyway.